c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+               "be accessed, so you'll have no way to view the data after "
+               "future replays of the owning graph.");
+  }
+
+  free_impl(it);
+}
+
+// Symmetric with NativeCachingAllocator::malloc for now,
+// although I don't think we absolutely need the symmetry.
+void malloc(void** devPtr, int device, size_t size, cudaStream_t stream) {
+  TORCH_INTERNAL_ASSERT(
+      0 <= device && static_cast<size_t>(device) < device_count,
+      "Invalid device index ",
+      device,
+      ": did you call init?");
+
+  // If stream is a null (default) stream,
+  // cudaMallocAsync infers the device from the ambient context,
+  // so we need to set the right ambient context.
+  CUDAGuard g(device);
+
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  lazy_init_device(device);
+
+  // Defensively checks for preexisting CUDA error state.
+  auto err = cudaGetLastError();
+  C10_CUDA_CHECK(err);
+
+  // TODO: Could we avoid calling cudaMallocAsync while holding general_mutex,
+  // perhaps by letting lazy_init_device use separate once_flags or an internal
+  // static initializer?
+  if (pytorch_used_bytes[device] + size > pytorch_memory_limits[device]) {
+    err = cudaErrorMemoryAllocation;
+  } else {
+    err = cudaMallocAsync(devPtr, size, stream);
+  }
+
+  if (err == cudaErrorMemoryAllocation) {
+    // Clears CUDA's internal error state so the user, if desired, can catch the OOM
+    // exception, free some stuff on the script side, and retry the allocation.
+    // This aligns with the behavior of alloc_block in CUDACachingAllocator.cpp.
+    cudaGetLastError();
+    size_t device_free;
+    size_t device_total;
+    C10_CUDA_CHECK(cudaMemGetInfo(&device_free, &device_total));
+    TORCH_CHECK_WITH(CUDAOutOfMemoryError,
+                     false,
+                     "Allocation on device ", device, " would exceed allowed memory.",
+                     "\nCurrently allocated     : ", format_size(pytorch_used_bytes[device]),
+                     "\nRequested               : ", format_size(size),
+                     "\nDevice limit            : ", format_size(device_total),
+                     "\nFree (according to CUDA): ", format_size(device_free),
+                     "\nPytorch limit (set by user-supplied memory fraction)"
+                     "\n                        : ", format_size(pytorch_memory_limits[device]));
+  } else {
+    C10_CUDA_CHECK(err);
+  }
+
+  auto inserted = ptr_info.emplace(*devPtr, PtrUsage(size, capture_underway));
+  TORCH_INTERNAL_ASSERT(inserted.second,
+                        "address returned by cudaMallocAsync already exists "
+                        "in ptr_info");
+
+  inserted.first->second.usage_streams.emplace_back(stream, device);
+
+  pytorch_used_bytes[device] += size;
+}
+
+} // anonymous namespace
+
+// Same pattern as CUDACachingAllocator.cpp.
+// Again, I don't think we absolutely need the symmetry,
+// but it's the simplest way to imitate the interface.
```

Comment:
This interface matters, because this allocator is how tensors know how to resize/free their tensor memory in contexts where they don't know how it was allocated. I'd recommend naming this CudaAsyncCachingAllocator. I wouldn't have created a NativeCachingAllocator; there is already a virtual indirection here so there is no need to ALSO do the boolean test one way or another.